grpc/server.go

```
@@ -0,0 +1,114 @@
+	ServerRegistration func(*grpc.Server)             // Callback for registering GRPC API Servers
+	StreamInterceptors []grpc.StreamServerInterceptor // A list of global GRPC stream interceptor functions to be called. Order is honored left to right.
+	UnaryInterceptors  []grpc.UnaryServerInterceptor  // A list of global GRPC unary interceptor functions to be called. Order is honored left to right.
+	CancelSignals      []os.Signal                    // OS Signals to be used to cancel running servers. Defaults to SIGINT/`os.Interrupt`.
+}
+
+// Server contains the configured GRPC server and related components
+type Server struct {
+	server        *grpc.Server
+	listenAddress string
+	cancelSignals []os.Signal // OS Signals to be used to cancel running servers. Defaults to SIGINT/`os.Interrupt`.
+}
+
+// NewDefaultConfig returns a default GRPC server config object. The caller must still supply the
+// name of the server and a serverRegistration callback. The serverRegistration callback is a
+// function that registers GRPC API servers with the provided GRPC server.
+func NewDefaultConfig(name string, serverRegistration func(*grpc.Server)) Config {
+	return Config{
+		Name:               name,
+		Address:            "127.0.0.1",
+		Port:               9111,
+		ServerRegistration: serverRegistration,
+		StreamInterceptors: []grpc.StreamServerInterceptor{},
+		UnaryInterceptors:  []grpc.UnaryServerInterceptor{},
+		CancelSignals:      []os.Signal{os.Interrupt},
+	}
+}
+
+// NewServer creates and returns a configured Server object given a GRPC configuration object.
+func (c Config) NewServer() (Server, error) {
+	server := grpc.NewServer(
+		grpc.StreamInterceptor(
+			grpc_middleware.ChainStreamServer(
+				c.StreamInterceptors...,
+			),
+		),
+		grpc.UnaryInterceptor(
+			grpc_middleware.ChainUnaryServer(
+				c.UnaryInterceptors...,
+			),
+		),
+	)
+	if c.ServerRegistration == nil {
+		return Server{}, fmt.Errorf("no server registration function provided")
```

I feel like letting the call to `ServerRegistration` panic instead of returning an error is acceptable since `ServerRegistration` being `nil` is a usage error on the part of the user of this library and I can't really think of a scenario where this would be a recoverable situation. Getting rid of the error from this method would enable a nice 1-liner `err := c.NewServer().Run()` 🤷‍♂